third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp

```diff
@@ -224,12 +224,101 @@ static inline Value truncAndCastFromInt(RewriterBase &rewriter, Location loc,
   return toVal;
 }
 
+// Permute lanes of the input val and apply reduction to permuted values.
+static Value permuteAndReduce(RewriterBase &rewriter, Location loc,
+                              StringRef intrinsic, Value val,
+                              Operation *reduxOp) {
+  Type valType = val.getType();
+  assert(valType.getIntOrFloatBitWidth() <= 32);
+
+  Type actualType = valType;
+  if (!valType.isInteger(32))
+    actualType = castToAndSExtInt(rewriter, loc, val, valType, 32);
+
+  auto b = TritonLLVMOpBuilder(loc, rewriter);
+  Value falseVal = b.false_val();
+  MLIRContext *ctx = rewriter.getContext();
+  Type retType = struct_ty({i32_ty, i32_ty});
+  Value perm =
+      LLVM::createLLVMIntrinsicCallOp(rewriter, loc, intrinsic, retType,
+                                      ValueRange{val, val, falseVal, falseVal})
+          ->getResult(0);
+  Value v0 = b.extract_val(i32_ty, perm, 0);
+  Value v1 = b.extract_val(i32_ty, perm, 1);
+
+  if (!valType.isInteger(32)) {
+    v0 = truncAndCastFromInt(rewriter, loc, v0, valType, 32);
+    v1 = truncAndCastFromInt(rewriter, loc, v1, valType, 32);
+  }
+  IRMapping mapping;
+  mapping.map(reduxOp->getOperand(0), v0);
+  mapping.map(reduxOp->getOperand(1), v1);
+  Value redx = rewriter.clone(*reduxOp, mapping)->getResult(0);
+  return redx;
+}
+
+// Apply warp reduction across lanes using llvm intrinsics in GFX950.
+// The input acc has the partial accumulated value from reduction within
+// threads. The output acc has the final accumulated value.
+//
+// The thread layout must be mfma input layout.
+// For the mfma_32x32 layout:
+//   step 1: use permlane32_swap() to swap the row 2 and 3 of acc and
+//           the row 0 and 1 of the copy of acc
+//   step 2: apply reduction to the result values to get final result
+// For the mfma_16x16 layout:
+//   step 1: use permlane32_swap() to swap the row 2 and 3 of acc and
+//           the row 0 and 1 of the copy of acc
+//   step 2: apply reduction to the result values to get the partial result
+//   step 3: use permlane16_swap() to swap the odd and even rows of
+//           the partial results
+//   step 4: apply reduction to get the final results
+static bool warpReduceCDNA4(RewriterBase &rewriter, Location loc,
+                            SmallVector<Value> &acc, triton::ReduceOp op) {
+  Operation *reduxOp = op.getSingleCombiner();
+  if (!reduxOp)
+    return false;
+
+  if (acc.size() != 1)
+    return false;
+  Value val = acc[0];
+  unsigned bits = val.getType().getIntOrFloatBitWidth();
+  if (bits > 32)
+    return false;
+
+  auto resultTy = op.getInputTypes()[0];
+  auto mfmaLayout = dyn_cast<AMDMfmaEncodingAttr>(resultTy.getEncoding());
+  if (!mfmaLayout)
```

Comment:
This is not enough. 
We not only need the srcLayout to be mfma, but also the reduction order is row order.

Reply:
I think we also want isTransposed=true for the mfmaLayout.

Reply:
If https://github.com/triton-lang/triton/pull/6594/files#r2059474101 makes sense to you, then we can generalize the condition for this special case, by 
- pass `numLaneToReduce` and `interleave` into warpReduceCDNA4
- mfma32Case = numLaneToReduce==2 && interleave==32
- mfma16Case = numLaneToReduce==4 && interleave==16
- if (!mfma32Case && !mfma16Case) return false;
- Then handle mfma32 and mfma16 case accordingly

Reply:
However, I'm still not convinced why acc.size() is 1.
The caller can pass a vector of acc's per thread to reduce. And you only need to add a for loop to process each acc in the same way.

Reply:
Right. I saw a for loop in warpReduce() of common codebase as well. I had thought the special mfma cases just had one acc per thread because that's the case for FA test case. But looks like it is not a general assumption.